nc/views.py

```diff
@@ -40,6 +41,37 @@ def search(request):
 class AgencyList(ListView):
     model = Agency
 
+    def get_context_data(self, **kwargs):
+        context = super(AgencyList, self).get_context_data(**kwargs)
+
+        if self.request.method == 'GET' and self.request.GET:
+            form = forms.AgencySearchForm(self.request.GET)
+            if form.is_valid():
+                agency = form.cleaned_data['agency']
+                return redirect('nc:agency-detail', agency.pk)
+        else:
+            form = forms.AgencySearchForm()
+
+        # Once we have the "letters present", we want to be able to iterate
+        # over categorized, sorted lists of agencies. Therefore we create
+        # a dict indexed by first letter.
+        sorted_agencies = {}
+
+        for agency in context['agency_list']:
+            initial = agency.name[:1]
+            if initial not in sorted_agencies:
+                sorted_agencies[initial] = []
+            sorted_agencies[initial].append(agency)
+
+        for key in sorted_agencies:
+            sorted_agencies[key].sort(key=lambda x: x.name)
+            sorted_agencies[key] = get_chunks(sorted_agencies[key])
+
+        sorted_agencies = sorted(sorted_agencies.items())
+
+        return dict(context, **{"sorted_agencies": sorted_agencies,
+                                "agency_form": form, })
```

Comment:
why faster?

my version does this twice: 
* make reference to object
* search dictionary to find key (don't find it)
* insert new key set to the new reference

your version does this:
* create new dictionary (the literal)
* perform the same steps as my version, but on the new dictionary
* munge new dictionary to kwargs format as it calls dict constructor
* make new dict
* walk through context duplicating all keys and setting them to new references to the objects in context
* for the kwargs, again perform the steps as my version, but on the 2nd new dictionary

(Huge caveats, largely that I've never read Python object code :) )